receiver/nvmlreceiver/client.go

```diff
@@ -0,0 +1,276 @@
+	nvmlCleanup = nil
+
+	defer func() {
+		// applicable to tagged releases of github.com/NVIDIA/go-nvml <= v0.11.6-0
+		if perr := recover(); perr != nil {
+			err = fmt.Errorf("%v", perr)
+		}
+	}()
+
+	ret := nvmlInit()
+	if ret != nvml.SUCCESS {
+		if ret == nvml.ERROR_LIBRARY_NOT_FOUND {
+			err = fmt.Errorf("libnvidia-ml.so not found")
+		} else {
+			err = fmt.Errorf("'%v'", nvml.ErrorString(ret))
+		}
+		return
+	}
+	logger.Sugar().Infof("Successfully initialized Nvidia Management Library")
+
+	nvmlCleanup = func() {
+		ret := nvml.Shutdown()
+		if ret != nvml.SUCCESS {
+			logger.Sugar().Infof("Unable to shutdown Nvidia Management library on '%v'", nvml.ErrorString(ret))
+		}
+	}
+
+	err = nil
+	return
+}
+
+func discoverDevices(logger *zap.Logger) ([]nvml.Device, []string, error) {
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		return nil, nil, fmt.Errorf("Unable to get Nvidia device count on '%v'", nvml.ErrorString(ret))
+	}
+
+	devices := make([]nvml.Device, count)
+	names := make([]string, count)
+	for i := 0; i < count; i++ {
+		devices[i], ret = nvml.DeviceGetHandleByIndex(i)
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get Nvidia device at index %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		uuid, ret := devices[i].GetUUID()
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get UUID of Nvidia device %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		names[i], ret = devices[i].GetName()
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get name of Nvidia device %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		logger.Sugar().Infof("Discovered Nvidia device %d of model %s with UUID %s", i, names[i], uuid)
+
+		currMode, _, ret := devices[i].GetMigMode()
+		if ret != nvml.SUCCESS {
+			logger.Sugar().Warnf("Unable to query MIG mode for Nvidia device %d", i)
+			continue
+		}
+		if currMode == nvml.DEVICE_MIG_ENABLE {
+			logger.Sugar().Warnf("Nvidia device %d has MIG enabled. GPU utilization queries may not be supported", i)
+		}
+	}
+
+	return devices, names, nil
+}
+
+func (client *nvmlClient) cleanup() {
+	if client.handleCleanup != nil {
+		client.handleCleanup()
+	}
+	if !client.disable {
+		client.logger.Info("Shutdown Nvidia Management Library client")
+	}
+}
+
+func (client *nvmlClient) getDeviceModelName(gpuID uint) string {
+	return client.devicesModelName[gpuID]
+}
+
+func (client *nvmlClient) collectDeviceMetrics() ([]nvmlMetric, error) {
+	// not strictly needed since len(client.devices) = 0; but, safer
+	if client.disable {
+		return nil, nil
+	}
+
+	deviceMetrics := client.collectDeviceUtilization()
+	deviceMetrics = append(deviceMetrics, client.collectDeviceMemoryInfo()...)
+	return deviceMetrics, nil
+}
+
+func (client *nvmlClient) collectDeviceUtilization() []nvmlMetric {
+	deviceMetrics := make([]nvmlMetric, 0, len(client.devices))
+
+	gpuUtil := nvmlMetric{name: "nvml.gpu.utilization"}
```

Comment:
Hm, I don't really want to create a new variable for the string constant or duplicate it. Since this is a bit of a nit, I think we can just leave it as-is.